src/browser-polyfill.js

```diff
@@ -396,36 +396,96 @@ if (typeof browser === "undefined") {
         // result, otherwise wait the promise related to the wrappedSendResponse
         // callback to resolve and send it as a response.
         if (isResultThenable) {
-          result.then(sendResponse, error => {
-            console.error(error);
-            // TODO: the error object is not serializable and so for now we just send
-            // `undefined`. Nevertheless, as being discussed in #97, this is not yet
-            // providing the expected behavior (the promise received from the sender should
-            // be rejected when the promise returned by the listener is being rejected).
-            sendResponse(undefined);
-          });
+          sendPromisedResult(result);
         } else {
-          sendResponsePromise.then(sendResponse, error => {
-            console.error(error);
-            // TODO: same as above, we are currently sending `undefined` in this scenario
-            // because the error oject is not serializable, but it is not yet the behavior
-            // that this scenario should present.
-            sendResponse(undefined);
-          });
+          sendPromisedResult(sendResponsePromise);
         }
 
         // Let Chrome know that the listener is replying.
         return true;
       };
     });
 
+    const wrappedSendMessageCallback = ({reject, resolve, unexpectedError}, reply) => {
+      if (reply && reply.__webExtensionPolyfillReject__) {
+        const {error} = reply;
+        // Convert the JSON representation of the error back to an Error instance,
+        // otherwise just reject with the error value as it is.
+        if (reply.__webExtensionPolyfillReject__.isErrorInstance) {
+          reject(new Error(error.message, error.fileName, error.lineNumber));
+        } else {
+          // If the original error wasn't an error instance, let's try to keep
+          // the behavior as close as possible to the Firefox behavior, e.g.
+          // print the error object as received in the console and then reject
+          // using a generic error object.
+          console.error(error);
+          reject(unexpectedError);
+        }
+      } else {
+        resolve(reply);
+      }
+    };
+
+    const wrappedSendMessage = (name, metadata, apiNamespaceObj, ...args) => {
+      if (args.length < metadata.minArgs) {
+        throw new Error(`Expected at least ${metadata.minArgs} ${pluralizeArguments(metadata.minArgs)} for ${name}(), got ${args.length}`);
+      }
+
+      if (args.length > metadata.maxArgs) {
+        throw new Error(`Expected at most ${metadata.maxArgs} ${pluralizeArguments(metadata.maxArgs)} for ${name}(), got ${args.length}`);
+      }
+
+      // Create an error object here, so that the original caller is part of the
+      // stack trace (similarly to the stack of the error raised on Firefox in
+      // this scenario).
+      const unexpectedError = new Error("An unexpected error occurred");
```

Comment:
I'm missing a test case that checks this behavior.

Reply:
This particular line is gone in the updated version.

A test case for the "An unexpected error occurred" is in the current set of tests but like I briefly explain in the inline comment, Firefox seems to present a different behavior in version 59 (where the error has the "An unexpected error occurred" message) and 60 (where it seems that it is currently "undefined"), and so the test cases is not currently asserting the error message (it just checks that an instance of Error is rejected in that case).